# Учебный проект 2_Анализ оттока клиентов банка

## Содержание

* [Описание проекта](#Описание)
* [Импорт библиотек Python и загрузка данных](#Импорт)
* [Предобработка данных](#Предобработка)
    * [Категоризация данных](#Категоризация)
* [Исследовательский анализ данных](#Исследование)
* [Вывод](#Вывод)

## Описание проекта <a class = 'anchor' id = 'Описание'></a>

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Датасет имеет следующую структуру:

* **children** - количество детей в семье;
* **days_employed** — общий трудовой стаж в днях;
* **dob_years** — возраст клиента в годах;
* **education** — уровень образования клиента;
* **education_id** — идентификатор уровня образования;
* **family_status** — семейное положение;
* **family_status_id** — идентификатор семейного положения;
* **gender** — пол клиента;
* **income_type** — тип занятости;
* **debt** — имел ли задолженность по возврату кредитов;
* **total_income** — ежемесячный доход;
* **purpose** — цель получения кредита;

## Импорт библиотек Python и загрузка данных <a class = 'anchorn' id = 'Импорт'></a>

Данный блок характеризуется следующими последовательными действиями:

1. Импорт библиотек Python:
    * для манипулирования данными;
    * для визуализации данных;
2. Загрузка данных в рабочую среду Jupyter Notebook. Инициализация переменной **data** для хранения загруженного датасета;
3. Вывод на экран параметров датасета:
    * Перечень столбцов набора данных;
    * Количество строк в наборе данных;
    * Количество столбцов в наборе данных.

In [4]:
# импорт библиотек Python
import pandas as pd #для анализа табличных данных
import numpy as np #для работы с объектами линейной алгебры

import matplotlib.pyplot as plt # для визуализации данных

In [5]:
# инициализация переменной 'data' и загрузка набора данных
data = pd.read_csv('~/Desktop/YandexPraktikum_projects/datasets/data.csv')

# вывод на экран перечня столбцов датасета
for column in data.columns.to_list():
    print('*', column)

print()
print('Количество строк в наборе данных:', data.shape[0])
print('Количество столбцов в наборе данных:', data.shape[1])

* children
* days_employed
* dob_years
* education
* education_id
* family_status
* family_status_id
* gender
* income_type
* debt
* total_income
* purpose

Количество строк в наборе данных: 21525
Количество столбцов в наборе данных: 12


**Вывод:**

1. Импортированы библиотеки Python:
    * для работы с данными табличного вида;
    * для визуализации данных.
2. Инициализирована переменная **data** и загружен рабочий датасет в рабочую среду;
3. Выведены на экран следующие параметры датасета:
    * перечень названий столбцов с данными;
    * количество строк в наборе данных;
    * количество столбцов в наборе данных.

## Предобработка данных <a class = 'anchor' id = 'Предобработка'></a>

Данный блок характеризуется следующими последовательными действиями:

1. Вывод на экран структуры и основной информации о датасете;
2. Проверка набора данных на пропущенные значения. Принятие решения об исключении пропущенных значений в датасете;
3. Проверка набора даных на наличие аномальных значений:
    * Проверка на аномально большие значения;
    * Проверка на аномально малые значения;
4. Изучение датасета на наличие дубликатов (явных и неявных). Выбор метода по исключению дубликатов из набора данных;
5. Категоризация данных. Создание новых сущностей в наборе данных по значениям из других столбцов;
6. Вывод по этапу.

In [6]:
# вывод первых 20 строчек на экран
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [7]:
# вывод основной информации о датасете на экран
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [8]:
# вывод на экран количества пустых значений в наборе данных
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [9]:
# вывод на экран доли пропущенных значений от общего количества данных в наборе
print('Доля пропущенных значений в наборе данных: {:.1%}'.format(data['days_employed'].isna().sum() / data.shape[0]))

Доля пропущенных значений в наборе данных: 10.1%


**Вывод по промежуточному этапу:**

Исходя из большого процента пропущенных значений, - **10.1%** - нельзя исключить строки, содержащие пустоты, из общего набора данных. Так можно исказить первоначальную информацию, хранящуюся в датасете.

Можно заполнить пустоты медианным значением, так как оно меньше подвержено влиянию выбросов, чем среднее арифметическое.

In [10]:
# заполнение пустых ячеек медианным значением по специализированным группам
income_type_ls = list(data['income_type'].unique())

for income_type in income_type_ls:
    median_value = data[data['income_type'] == income_type]['total_income'].median()
    data.loc[(data['income_type'] == income_type) & (data['total_income'].isna()), 'total_income'] = median_value

In [11]:
# вывод на экран описательной статистики по набору данных
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.652253e+05  
std            1.420324      0.272661  9.804367e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.425944e+05  
75%            1.000000      0.000000  1.955499e+05  
max            4.000000      1.000000  2.265604e+06

**Вывод по промежуточному этапу:**

В наборе данных присутствуют аномальные значения:

    'children' - количество детей у клиентов: -1 и 20;
    'day_employed' - отрицательные значения трудового стажа в днях

Скорее всего значения отрицательные значения трудового стажа - техническая ошибка при заполнении базы данных. Количество отрицательных значений огромно - 3 квартиль (75-перцентиль) находится так же в отрицательной зоне. По этой причине нельзя удалить из датасета эти строки.

Принято решение заменить отрицательные значения на положительные.

In [12]:
# меняем отрицательные значения на положительные
data['days_employed'] = abs(data['days_employed'])

In [13]:
# вывод медианного значения трудового стажа в днях для каждого типа занятости
data.groupby('income_type')['days_employed'].median()

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

**Вывод по промежуточному этапу:**

У двух типов (безработные и пенсионеры) получаются аномально большие значения.

Исправление таких значений должно быть основано либо на четком понимании контекста исходного набора данных, либо на экспертном мнении руководителя, либо на вводной от заказчика.

Исправление данных значений выходит за рамки выполнения учебного проекта, поэтому данные группы останутся без изменений.

In [14]:
# вывод на экран уникальных значений столбца 'children'
data.children.unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [15]:
# вывод на экран количество строк, содержащих аномальные отрицательные значения
print('Количество клиентов с \"отрицательным\" значением детей:', data[data['children'] < 0]['children'].sum())

Количество клиентов с "отрицательным" значением детей: -47


In [16]:
# вывод на экран количество строк, содержащих аномальные положительные значения
print('Количество клиентов с аномально большим количеством детей:', data[data['children'] == 20]['children'].sum())

Количество клиентов с аномально большим количеством детей: 1520


**Вывод по промежуточному этапу:**

Количество записей с аномальными показателями количества детей ничтожно мало по сравнению с количеством строк общего набора данных.

Исходя из этого, строки, содержащие такие значения можно удалить.

In [17]:
# удаление строк с аномальными данными о количестве детей
data = data[(data['children'] >= 0) & (data['children'] < 20)]

# проверка удаления строк
data.children.unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

In [18]:
# заполнение пропусков в столбце 'days_employed' медианными значениями по каждому типу занятости
for income_type in income_type_ls:
    median_value = data[data['income_type'] == income_type]['days_employed'].median()
    data.loc[(data['income_type'] == income_type) & (data['days_employed'].isna()), 'days_employed'] = median_value

# проверка заполнения пропусков
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [19]:
# замена вещественного типа данных на целочисленный
data['total_income'] = data['total_income'].astype('int')

In [20]:
# понижение регистра в столбце 'education'
data['education'] = data['education'].str.lower()

# вывод на экран количество строк-дубликатов
data.duplicated().sum()

71

In [21]:
# удаление дубликатов из набора данных
data = data.drop_duplicates()

### Категоризация данных <a class = 'anchor' id = 'Категоризация'></a>

На основании приведенных диапазонов необходимо создать новый столбец, содержащий категории данных - `total_income_category`:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [22]:
# инициализация функции categorize_income()
def categorize_income(total_income):
    if total_income < 30001:
        return "E"
    elif total_income < 50001:
        return "D"
    elif total_income < 200001:
        return "C"
    elif total_income < 1000001:
        return "B"
    else:
        return "A"

In [23]:
# создание столбца 'total_income_category'
data['total_income_category'] = data['total_income'].apply(categorize_income)

На основании данных, содержащихся в столбце `purpose` (цели взятия кредита) необходимо создать новый столбец, содержащий категории данных - `purpose_category`:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [24]:
# перечень уникальных целей взятия кредита
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [25]:
# инициализация функции categorize_purpose()
def categorize_purpose(purpose):
    if 'авто' in purpose:
        return 'операции с автомобилем'
    elif 'обр' in purpose:
        return 'получение образования'
    elif 'свадь' in purpose:
        return 'проведение свадьбы'
    else:
        return 'операции с недвижимостью'

In [26]:
# создание столбца 'purpose_category'
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Вывод:**

1. На экран выведена структура набора данных, а также основная информация о датасете: идентифицированы столбцы, содержащие пропуски; изучены типы значений в столбцах;

2. Набор проверен на пропущенные значения. Пустые ячейки обнаружены в следующих столбцах:

* 'days_employed' - 2 174 значения. Что составляет **10.1%** от общего количества строк;
* 'total_income' - 2 174 значения. Что составляет **10.1%** от общего количества строк;

Среди причин появления могут быть технические ошибки при формировании набора данных, а также человеческий фактор - ошибки при внесении данных в базу, а также сам факт не указания информации клиентом о себе.

Используя вводную из описания задания, пустые значения в столбце 'total_income' были заполнены медианным значением по каждой группе клиентов.

Заполнение пустых ячеек медианным значением приоритетнее среднего по причине того, что медиана, как статистическая метрика, меньше подвержена влиянию аномальных значений.

3. Набор данных проверен на аномальные значения - обнаружены в столбцах 'children' и 'days_employed'. Аномальные значения устранены путем приведения к положительному знаку (столбец 'days_employed') и исключению строк из общего набора данных (столбец 'children');

4. Устранены явные и неявные дубликаты - путем понижения регистра и удаления строк, представляющих собой задублированные значения;

5.Выполнена категоризация данных - созданы столбцы 'total_income_category' и 'purpose_category'.

**Набор данных подготовлен к дальнейшему проведению исследования.**


## Исследовательский анализ данных

Данный блок характеризуется следующими последовательными действиями:

1. Изучение зависимости между количеством детей и возвратом кредита в срок;
2. Изучение зависимости между семейным положением и возвратом кредита в срок;
3. Изучение зависимости между уровнем дохода и возвратом кредита в срок;
4. Изучение зависимости между целями кредита и возвратом кредита в срок;
5. Вывод по итогу этапа.

In [36]:
# инициализация пользовательской функции по созданию сводной таблицы
def create_pivot(data : pd.DataFrame, feature : str):
    pivot_data = (data.groupby(feature).agg({'debt' : 'sum',
                                             'dob_years' : 'count'})
                        .rename(columns = {'debt' : 'debtors',
                                           'dob_years' : 'all_clients'}))
    pivot_data['share_debtors'] = round(pivot_data['debtors'] * 100 / pivot_data['all_clients'], 2)
    return pivot_data

In [37]:
# формирование свода по зависимости возврата кредита от количества детей
create_pivot(data, 'children')

debtors  all_clients  share_debtors
children                                     
0            1063        14091           7.54
1             444         4808           9.23
2             194         2052           9.45
3              27          330           8.18
4               4           41           9.76
5               0            9           0.00

**Вывод по промежуточному этапу:**

По содержимому сводной таблицы можно сделать предварительный вывод о том, что взаимосвязи между определенным количеством детей и наличии задолженности - **нет**. Процент должников равномерно распределен по всем категориям клиентов в зависимости от определенного количества детей.

Только у одной группы - те клиенты, у кого 5 детей - нет задолженности. Данная группа не является отражением общей картины, позволяющей сделать вывод о определенной взаимосвязи долга перед банком и количестве детей.

In [38]:
# формирование свода по зависимости возврата кредита от семейного положения
create_pivot(data, 'family_status')

debtors  all_clients  share_debtors
family_status                                             
Не женат / не замужем      273         2796           9.76
в разводе                   84         1189           7.06
вдовец / вдова              63          951           6.62
гражданский брак           385         4134           9.31
женат / замужем            927        12261           7.56

**Вывод по промежуточному этапу:**

В данной таблице уже более выраженная ситуация: больший процент должников имеют группы **"Не женат/не замужем"** и **"Гражданский брак"**.

В остальных группах низкий процент должников объясняется некоторыми правовыми и гражданскими издержками:
1. С кредиторов могут быть списаны долги перед банком, если кто-то из семейной пары скончался - этим объясняется низкий процент в группе "вдовец/вдова";
2. В группе "женат/замужем" низкий процент должников объясняется ведением общего домового хозяйства и совместным заработком, поэтому и сама потребность в кредите отпадает;
3. Вполне вероятно, что в должниках - в группе "в разводе" - оказались те люди, на которых суд наложил право выплатить кредит, взятый в семейной паре. Объяснить низкий процент в данной группе достаточно сложно - необходимо владеть более полной картиной ситуации.

In [39]:
# отдельно сформируем свод по семейному положению и уровню дохода
famIncome_pivot = (data.groupby(['family_status', 'total_income_category'])
                   .agg({'debt' : 'sum', 'dob_years' : 'count'})
                   .rename(columns = {'debt' : 'debtors', 'dob_years' : 'all_clients'}))
famIncome_pivot['share_debtors'] = round(famIncome_pivot['debtors'] * 100 / famIncome_pivot['all_clients'], 2)

famIncome_pivot

debtors  all_clients  \
family_status         total_income_category                         
Не женат / не замужем A                            0            2   
                      B                           62          667   
                      C                          209         2100   
                      D                            2           25   
                      E                            0            2   
в разводе             A                            0            3   
                      B                           18          282   
                      C                           64          887   
                      D                            2           17   
вдовец / вдова        B                            9          154   
                      C                           54          775   
                      D                            0           22   
гражданский брак      A                            1            6   
                      B                           64          973   
                      C                          315         3087   
                      D                            5           62   
                      E                            0            6   
женат / замужем       A                            1           14   
                      B                          201         2938   
                      C                          711         9072   
                      D                           12          223   
                      E                            2           14   

                                             share_debtors  
family_status         total_income_category                 
Не женат / не замужем A                               0.00  
                      B                               9.30  
                      C                               9.95  
                      D                               8.00  
                      E                               0.00  
в разводе             A                               0.00  
                      B                               6.38  
                      C                               7.22  
                      D                              11.76  
вдовец / вдова        B                               5.84  
                      C                               6.97  
                      D                               0.00  
гражданский брак      A                              16.67  
                      B                               6.58  
                      C                              10.20  
                      D                               8.06  
                      E                               0.00  
женат / замужем       A                               7.14  
                      B                               6.84  
                      C                               7.84  
                      D                               5.38  
                      E                              14.29

**Вывод по промежуточному этапу:**

В данном разрезе уже интересно посмотреть на высокие доли должников в группах "Не женат/не замужем" и "Гражданский брак".

В этих группах семейного положения самый большой процент должников имеет **"средний" доход - категория "C"**.

Вполне вероятно, что данная категория клиентов имеет крупные цели взятия кредита, - покупка машины или квартиры - которые самостоятельно "потянуть" не смогли. А гражданский брак подразумевает, что люди все равно самостоятельно отвечают за взятые на себя обязательства.

Интересно в дальнейшем посмотреть на то, как взаимосвязаны группы доходов и цели взятия кредита.

In [40]:
# формирование свода по зависимости возврата кредита от категории дохода
create_pivot(data, 'total_income_category')

debtors  all_clients  share_debtors
total_income_category                                     
A                            2           25           8.00
B                          354         5014           7.06
C                         1353        15921           8.50
D                           21          349           6.02
E                            2           22           9.09

**Вывод по промежуточному этапу:**

По содержимому сводной таблицы можно сделать предварительный вывод о том, что люди с категорией заработка **С** больше склонны оставаться в статусе "должник", чем остальные категории клиентов.

Как следует и из предыдущего исследования - вполне вероятно, что эти клиенты определили для себя дорогостоящие цели кредита, которые не смогли оплатить в срок.

In [41]:
# формирование свода по зависимости возврата кредита от цели кредита
create_pivot(data, 'purpose_category')

debtors  all_clients  share_debtors
purpose_category                                             
операции с автомобилем        400         4279           9.35
операции с недвижимостью      780        10751           7.26
получение образования         369         3988           9.25
проведение свадьбы            183         2313           7.91

**Вывод по промежуточному этапу:**

В данном случае есть две ярко выраженные цели взятия кредита, где больше всего должников: **"Операции с автомобилем" и "Получение образования"**.

Необходимо совместить свод по целям взятия кредита и уровнем заработка клиентов, чтобы составить более полную картину.

In [42]:
# отдельно сформируем свод по цели взятия кредита и уровню дохода
purposeIncome_pivot = (data.groupby(['purpose_category', 'total_income_category'])
                   .agg({'debt' : 'sum', 'dob_years' : 'count'})
                   .rename(columns = {'debt' : 'debtors', 'dob_years' : 'all_clients'}))
purposeIncome_pivot['share_debtors'] = round(purposeIncome_pivot['debtors'] * 100 / purposeIncome_pivot['all_clients'], 2)

purposeIncome_pivot

debtors  all_clients  \
purpose_category         total_income_category                         
операции с автомобилем   A                            0            2   
                         B                           85         1022   
                         C                          305         3181   
                         D                           10           72   
                         E                            0            2   
операции с недвижимостью A                            1           17   
                         B                          172         2558   
                         C                          599         7997   
                         D                            6          168   
                         E                            2           11   
получение образования    A                            1            4   
                         B                           69          890   
                         C                          297         3014   
                         D                            2           75   
                         E                            0            5   
проведение свадьбы       A                            0            2   
                         B                           28          544   
                         C                          152         1729   
                         D                            3           34   
                         E                            0            4   

                                                share_debtors  
purpose_category         total_income_category                 
операции с автомобилем   A                               0.00  
                         B                               8.32  
                         C                               9.59  
                         D                              13.89  
                         E                               0.00  
операции с недвижимостью A                               5.88  
                         B                               6.72  
                         C                               7.49  
                         D                               3.57  
                         E                              18.18  
получение образования    A                              25.00  
                         B                               7.75  
                         C                               9.85  
                         D                               2.67  
                         E                               0.00  
проведение свадьбы       A                               0.00  
                         B                               5.15  
                         C                               8.79  
                         D                               8.82  
                         E                               0.00

**Вывод по промежуточному этапу:**

Данный свод дает более полную информацию. Как и предполагалось большая часть должников в двух группах "операции с автомобилем" и "получение образования" имеют средний уровень дохода - категория "С".

Стоит так же отметить, что и в других группах, сформированных по целям взятия кредита, в должниках оказались именно клиенты со средним уровнем заработка.

In [43]:
# отдельно сформируем свод по цели взятия кредита и количеству детей
childPurpose_pivot = (data.groupby(['purpose_category', 'children'])
                   .agg({'debt' : 'sum', 'dob_years' : 'count'})
                   .rename(columns = {'debt' : 'debtors', 'dob_years' : 'all_clients'}))
childPurpose_pivot['share_debtors'] = round(childPurpose_pivot['debtors'] * 100 / childPurpose_pivot['all_clients'], 2)

childPurpose_pivot

debtors  all_clients  share_debtors
purpose_category         children                                     
операции с автомобилем   0             243         2845           8.54
                         1             103          963          10.70
                         2              48          399          12.03
                         3               5           60           8.33
                         4               1           10          10.00
                         5               0            2           0.00
операции с недвижимостью 0             476         7074           6.73
                         1             200         2447           8.17
                         2              88         1035           8.50
                         3              13          169           7.69
                         4               3           21          14.29
                         5               0            5           0.00
получение образования    0             229         2642           8.67
                         1              90          866          10.39
                         2              46          403          11.41
                         3               4           69           5.80
                         4               0            7           0.00
                         5               0            1           0.00
проведение свадьбы       0             115         1530           7.52
                         1              51          532           9.59
                         2              12          215           5.58
                         3               5           32          15.62
                         4               0            3           0.00
                         5               0            1           0.00

**Вывод по промежуточному этапу:**

Изучая данный свод, можно сделать вывод о том, что большая часть должников в категориях "получение образования" и "операции с автомобилем" имеет детей и, притом, не одного.

## Общий вывод по исследованию <a class = 'anchor' id = 'Вывод'></a>

После проведения исследования можно сформировать выводы о портрете должника банка:
1. Уровень дохода клиента находится в категории "С";
2. Целью кредита клиента являются или "операции с автомобилем", или "получение образования";
3. Семейное положение "не женат/не замужем" или "гражданский брак"

Скорее всего при совпадении указанных признаков алгоритм бинарной классификации, построенный для банка, выдаст решение "отказать в выдаче кредита" новому клиенту банка.